### PR TITLE
Fix/286 - Support Eslint 8.x

### LINKED ISSUE #286: Support ESLint 8.x
ESLint v8.0.0 is [released](https://eslint.org/blog/2021/10/eslint-v8.0.0-released) 🎉

It would be awesome to have official ESLint 8 support. 👊
I'm happy to help where I can of course 🙂

### PR DESCRIPTION
**Pull request by gboer**
Fixes #286 

I had to make some more updates to other packages to make things actually work again, but I guess that never really hurts.

### CODE REVIEW

**Comment by nanpx** [comment]
Hi @vilchik-elena is there a rough estimate on getting this merged and released? Is there anything I can do to help? Thanks!

**Comment by Uzlopak** [comment]
We need this :)

**Comment by Uzlopak** [comment]
We need this :)

**Comment by nanpx** [comment]
Hi @yassin-kammoun-sonarsource @andrea-guarino-sonarsource @saberduck 

I'm reaching out as you all are listed as top contributors to this repository. The community has a large need to support ESLint v8, but we're not getting much response on the PR from SonarSource. Can you all help us determine next steps to get this merged and released? Thank you!

File: ruling/index.ts [hunk 1]
```diff
@@ -47,7 +47,7 @@ function run() {
   console.log("");
 
   const cli = new CLIEngine({
```
Comment (aarongoldenthal) [inline ruling/index.ts:49]: In `eslint` 8 the [`CLIEngine` class has been removed](https://eslint.org/docs/user-guide/migrating-to-8.0.0#the-cliengine-class-has-been-removed), but the migration guide has details on the required changes.

**Comment by gboer** [comment]
Thanks @aarongoldenthal , you're right indeed. Unfortunately this uncovered a whole host of issues everywhere. This PR will be a huge one when it's finally done..

**Comment by nanpx** [comment]
> Thanks @aarongoldenthal , you're right indeed. I only ran the tests and that passed without issues. But that didn't run the 'ruling' target, which did fail for eslint 8. Unfortunately fixing this, uncovered a whole host of issues everywhere. This PR will be a huge one when it's finally done..

@gboer feel free to start with my first attempt - https://github.com/nanpx/eslint-plugin-sonarjs/commit/a5ce538a29e0cf7a714abe46246921f3bac50ea1. I made it pretty close, just couldn't quite figure out some of the typescript things (see my comment here - https://github.com/SonarSource/eslint-plugin-sonarjs/issues/286#issuecomment-946086241)

I accounted for the breaking changes with the CLI there, as well as the breaking changes in the type definitions. Feel free to ping me if you need help with anything.

**Comment by gboer** [comment]
Thanks @nanpx :) I see that I've already solved all of the issues you found. Not sure what that error is that you got, I still haven't seen that yet. The biggest issue I found, was that the types of eslint are overridden and hid a lot of issues. I threw them out and now fixing all inconsistencies between the incorrect overridden types and what it should be. Making progress, but it's slow going unfortunately.

**Comment by gboer** [comment]
Well, @nanpx , I found your issue as well. Fixed it indeed by adding `requireConfigFile: false` to the configuration, but now I'm stuck with errors that you quite likely saw as well. This for instance: `Parsing error: 'return' outside of function. (6:2)`. The idea is that you add `globalReturn: true` to `ecmaFeatures`, but eslint overrides it and always sets it to `false`. But even changing eslint itself to set it to `true`, I get the next error `Parsing error: Missing semicolon. (23:4)`. It's just one thing after another and it is all due to the Babel upgrade. Haven't figure out a way to fix this yet unfortunately.

**Comment by gboer** [comment]
Alright, I gave up on upgrading babel-eslint. I have no idea how to make all of the 'ruling test code' work with just one Babel configuration. And even if you use different configurations for different projects within the ruling test code, you still run into parsing issues that I haven't been able to resolve. Especially React it does not like at all.

What is there now works. The only thing that should really be done, is to stop using @typescript-eslint/experimental-utils typings. All code should switch to @types/eslint instead, because that's actually in line with eslint itself. But I was starting to change it and that means another couple days worth of work. Something I'm not really up for right now.

File: tests/rules/no-empty-collection.test.ts [hunk 2]
```diff
@@ -247,7 +253,32 @@ ruleTester.run('Empty collections should not be accessed or iterated', rule, {
               console.log(array[2]);
               console.log(array[2]);
             }`,
-      errors: 4,
```
Comment (gboer) [inline tests/rules/no-empty-collection.test.ts:250]: it's not possible anymore to define a number as the number of errors. You can only specify the entire error itself.

File: typings/eslint.d.ts [hunk 3]
```diff
@@ -1,28 +0,0 @@
-/*
```
Comment (gboer) [inline typings/eslint.d.ts:1]: using @types/eslint now instead

File: src/utils/types.ts [hunk 4]
```diff
@@ -1,92 +0,0 @@
-/*
```
Comment (gboer) [inline src/utils/types.ts:1]: all of these mangled types are conflicting with the matching types of @types/eslint or @typescript-eslint/experimental-utils, so I replaced all of them with the types coming from those packages instead.

**Comment by vilchik-elena** [comment]
And https://github.com/SonarSource/eslint-plugin-sonarjs/pull/302 is finally merged 🥳 
I will go for the release now

**Comment by vilchik-elena** [comment]
Thanks @gboer!

**Comment by gboer** [comment]
Nice! Thanks for merging @vilchik-elena  :)

**Pull request closed without merging**